### PR TITLE
Database mapping macro

### PR DESCRIPTION
**Pull request by spalladino**
Mapping macro, similar to YAML or JSON mapping, to populate an instance of a class from a resultset.

### CODE REVIEW

**Comment by bcardiff** [comment]
This depends on Hash been ordered, which may change in the near future https://github.com/crystal-lang/crystal/issues/2331

**Comment by spalladino** [comment]
I don't see where it depends on the hash being sorted, as it matches the column name with the mapping key (exactly like the YAML mapping does)

**Comment by spalladino** [comment]
As for the travis check failure, it seems to be a change in Crystal 0.14 (it was working fine in 13)

**Comment by bcardiff** [comment]
I meant that https://github.com/bcardiff/crystal-db/pull/2/files#diff-89ed3becdd76dae7d10377988fac60a5R91 iterates the fields in certain order.

**Comment by bcardiff** [comment]
@spalladino could you rebase & update for the removal of `read?`. Maybe the last changes in JSON mapping to support nillable/unions offers some guidance to deal better with unions.

File: src/db/mapping.cr [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                  %rs.read(Union({{value[:type]}} | Nil))
+                {% else %}
+                  %rs.read({{value[:type]}})
+                {% end %}
+          {% end %}
+          else
+            {% if strict %}
+              raise DB::MappingException.new("unknown result set attribute: #{col_name}")
+            {% else %}
+              %rs.read(Nil)
```
Comment (bcardiff) [inline src/db/mapping.cr]: %rs.read  (without arguments) maybe ?

File: src/db/mapping.cr [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      def {{key.id}}=(_{{key.id}} : {{value[:type]}} {{ (value[:nilable] ? "?" : "").id }})
+        @{{key.id}} = _{{key.id}}
+      end
+
+      def {{key.id}}
+        @{{key.id}}
+      end
+    {% end %}
+
+    def self.from_rs(%rs : DB::ResultSet)
```
Comment (bcardiff) [inline src/db/mapping.cr]: this should go away, we can merge it as is, and remove it when adapting to the new query_one/query_all api.
agreed?

Comment (spalladino) [inline src/db/mapping.cr]: I'd like to keep this to be consistent with the JSON and YAML mapping APIs

**Pull request merged**